docs/src/source_gen.md

```diff
@@ -0,0 +1,481 @@
+end
+```
+
+```mermaid
+flowchart TD
+    subgraph PriorsAndParameters
+        S4["(S4) mu[k] ~ dnorm"]
+        S5["(S5) tau[k] ~ dgamma"]
+        S6["(S6) transition[k, 1:K] ~ ddirch"]
+        S7["(S7) pi[1:K] ~ ddirch"]
+    end
+
+    subgraph InitialState
+        S3["(S3) s[1] ~ dcat(pi[1:K])"]
+    end
+
+    subgraph MainLoop [for i in 1:T]
+        S1["(S1) s[i] ~ dcat(transition[s[i-1], 1:K])"]
+        S2["(S2) Y[i] ~ dnorm(mu[s[i]], tau[s[i]])"]
+    end
+
+    S7 --> S3
+    S6 --> S1
+    S3 --> S1
+    S1 -- loop-carried --> S1
+    S4 --> S2
+    S5 --> S2
+    S1 --> S2
+
+    %% Note: alpha is assumed to be an input/hyperparameter
+```
+
+In this example, there is a self loop on `S1`. These represent the state transition.
+To see that we can actually sequentially execute the program, we can compute the dependence vectors.
+
+The computation of the dependence vectors is done in the following steps:
+When executing the for loop from `i = 2` to `i = T`, we need to compute the dependence vector. Let's examine the case when `i = 2`:
+
+- LHS: `s[2]` is defined at iteration `i=2`
+- RHS: `s[1]` is defined at iteration `i=1`
+- Dependence vector: $\vec{d} = \vec{j}_{sink} - \vec{i}_{source} = (2) - (1) = (1)$
+
+All subsequent iterations follow the same pattern and have the same dependence vector of $(1)$. Because all dependence vectors are lexicographically non-negative, the loop is sequentially valid.
+
+This requires storing the loop variable `i` for each variable, but we already computed this with JuliaBUGS compilation, so not much overhead is required. 
+
+It should be noted that this approach doesn't scale well for general Julia programs, but it works appropriately for BUGS since the compilation process already has a time complexity of $O(N)$ with respect to the number of variables.
+
+It is worth pause here and give a summary:
+    
+To transform a BUGS program into a sequentially valid program, we will only apply two simple transformations, loop fission and statement reordering. These transformations will not need to modify loop bounds, renaming variables, adding any control flow, or make any changes to specific statements.
+
+A program that can be transformed are determined by the following procedure:
+
+First a statement dependence graph is computed.
+
+If the statement dependence graph is acyclic, then we topologically sort the statements and fission all the loops to create a sequentially valid program.
+
+In case the statement dependence graph is not acyclic.
+If the statements that form a cycle are all from the same loop (potentially at different nested levels), then we compute the dependence vectors to determine if the loop is sequentially valid. 
+
+Otherwise, the program need to be rewritten.
+
+We don't attempt to apply further transformations to the program, because it is a hard problem. And to give a flavor of the difficulty of rewriting programs, let's consider the following example:
```

Comment:
```suggestion
We don't attempt to apply further transformations to the program, because it is a hard problem. We will use the following example to show why program transformations can be a difficult task. We will not attempt to implement the transformation demonstrated here.

Consider this model,
```